Make Profile-screen use Mastodon.Entity.Account instead of MastodonUser

zeitschlag opened the pull request:
- Make Profile-screen use Mastodon.Entity.Account instead of MastodonUser
- Fix things left and right
- Just like with the Status-refactoring I'm expecting more bugs to show up over time.

<details>
<summary>Checklist</summary>
- [x] Sharing/Blocking/Muting/Domain Block
- [x] Follow/Unfollow, Request
- [x] Being Followed/Unfollowed
- [x] Polls
- [x] Notification (Avatar-Buttons, auch bei Quoted Posts)
- [x] Show/Hide sensitive content
- [x] Profile: Check Relationship-Button
- [x] Reporting
- [x] Compose/Answer
- [x] Retoot, Favorite, Delete
- [ ] ~~Widgets!~~ See #1234
- [x] ShareExtension
- [x] StatusEdits
</details>

kimar committed:
Remove usage of Status (IOS-176)

kimar committed:
Fix loading of home timeline (IOS-176)

kimar committed:
Fix notifications timeline loading (IOS-176)

kimar committed:
Merge branch 'develop' into remove_status

kimar committed:
Fix threads not working (IOS-176)

kimar committed:
Fix favorite action (IOS-176)

zeitschlag committed:
Sprinkle in some deprecations to see what needs to be done with MastodonUser (IOS-192, IOS-189)

kimar committed:
Reimplement favorite/reblog state (IOS-176)

kimar committed:
Merge branch 'develop' into remove_status

kimar committed:
Fix handling of (re-)reblogged and favorites state in home timeline (IOS-176)

kimar committed:
Fix manual refresh of home timeline not working (IOS-176)

kimar committed:
Fix duplicate entry after reblog (IOS-176)

kimar committed:
Reimplement timeline anchoring (IOS-176)

kimar committed:
Merge branch 'develop' into remove_status

kimar committed:
Fix tabbar avatar not visible (IOS-176)

kimar committed:
Don't try to resolve self-relationship

kimar committed:
Fix translate not working (IOS-176)

kimar committed:
Merge branch 'develop' into remove_status

kimar committed:
Fix compile issues after resolving merge conflicts

kimar committed:
Fix CW cannot be hidden (IOS-176)

kimar committed:
Fix bookmarked state (IOS-176)

kimar committed:
Fix missing case in StatusFetchedResultsController

kimar committed:
Improve StatusFetchedResultsController update handling (IOS-176)

kimar committed:
Improve flickering when interacting with posts (IOS-176)

kimar committed:
Preserve isSensitiveToggled state when interacting with posts (IOS-176)

kimar committed:
Fix sensitive toggled state for reblogs (IOS-176)

kimar committed:
Merge branch 'develop' into remove_status

kimar committed:
Fix user profile not loaded from status on home timeline (IOS-176)

kimar committed:
Implement caching for home timeline items (IOS-176)

kimar committed:
Fix status eventuall nil MastodonNotification

kimar committed:
Fix notifications show empty view (IOS-198)

kimar committed:
Implement cache for Notifications (IOS-200)

kimar committed:
Fix eventually incorrect domain used

kimar committed:
Fix missing poll information (IOS-201)

kimar committed:
Fix non-ui-thread modification in when editing poll (IOS-203)

kimar committed:
Remove unused code

kimar committed:
Fix Poll appearance / layout / issues (IOS-201)

kimar committed:
Fix status deletion not updating data model (IOS-205)

zeitschlag committed:
Fix Documentation-URL

zeitschlag committed:
WIP: Comment out and replace user with status (IOS-192)

kimar committed:
Fix favorited post might lose repost header (IOS-206)

kimar committed:
Remove status persistence (IOS-176)

zeitschlag committed:
Use placeholder for me (IOS-192)

kimar committed:
Fix CW layout issues (IOS-207)

zeitschlag committed:
Remove me/remote-profile-viewmodel (IOS-192)

zeitschlag committed:
Fix fields (IOS-192)

zeitschlag committed:
Show header-meta-information about account (IOS-192)

kimar committed:
Fix status thread CW not shown and interacted status not updated (IOS-208, IOS-210)

kimar committed:
Fix issue caused by duplicate IDs after reloadingtimeline

zeitschlag committed:
Add JSON-based account-persistence (IOS-192)

This is per user.

Also: Fetch authenticated accounts regularly
Also: Move Persistence-stuff to MastodonCore because.

kimar committed:
Implement thread reload after publishing reply (IOS-176)

zeitschlag committed:
Remove placeholder (IOS-192)

zeitschlag committed:
Replace MastodonUser.ID with string (IOS-192)

kimar committed:
Update datasource on delete statuses in timelines and thread (IOS-212)

kimar committed:
Update posts on timeline and thread if edited (IOS-211)

zeitschlag committed:
Migrate UserList to use Accounts (IOS-192)

zeitschlag committed:
Remove user from UserItem-lists (IOS-192)

Also: Remove CoreData and replace User with Account in several places

zeitschlag committed:
Try to fix domain (IOS-192)

zeitschlag committed:
Fix domain :facepalm: (IOS-192)

zeitschlag committed:
Add relationship to ProfileViewModel (IOS-192)

kimar committed:
Fix Polls not showing up (IOS-213)

zeitschlag committed:
Persist user after login (IOS-192)

zeitschlag committed:
Remove obsolete Core Data-controller

zeitschlag committed:
Configure barbutton-items on profile-screen (IOS-192)

zeitschlag committed:
Re-enable "Follows you"-indicator (IOS-192)

zeitschlag committed:
Re-enable blur-effect for pictures (IOS-192)

zeitschlag committed:
Refactor Mute/Block/HideReblogs to work with relationships (IOS-192)

Also: Connect Blocking/BlockedBy

zeitschlag committed:
Kill some forced unwrapping and hide bio when blocked/blocking (IOS-192)

First rule of optionals: Don't use forced unwrapping!

zeitschlag committed:
Reenable followers/followees (IOS-192)

zeitschlag committed:
Configure Profile-button based on relationship and accounts (IOS-192)

Also `me` is not optional anymore as we need it

zeitschlag committed:
Show correct account for "quoted" posts in notifications (IOS-192)

kimar committed:
Fix DataSourceFacade+Bookmark protocol constraints (IOS-176)

kimar committed:
Tie JSON cache to domain based unique identifier

kimar committed:
Remove unused code

kimar committed:
Fix incorrectly named variable

kimar committed:
Remove switch case

kimar committed:
Implement alternative Author for DataSourceFacade.MenuContext (IOS-176)

zeitschlag committed:
Show text on relationship-button based on relationship (IOS-192)

zeitschlag committed:
Make image clickable again (IOS-192)

zeitschlag committed:
Fix bug in follows-you-label (IOS-192)

:facepalm: RTFM (and add some docs)

zeitschlag committed:
Add some actions to Relationship-action-button (IOS-192)

Well, it's basically just unblock, unmute or follow/unfollow

kimar committed:
Address PR Feedback (IOS-176)

zeitschlag committed:
Move directory-stuff into its own FileManager-file (IOS-192)

zeitschlag committed:
User UserIdentification for search and accounts (IOS-192)

Thanks to @kimar!

zeitschlag committed:
Remove user (IOS-192)

zeitschlag committed:
Fix warnings (IOS-192)

zeitschlag committed:
Re-enable pending-check (IOS-192)

zeitschlag committed:
We don't use the author when composing toots (IOS-192)

zeitschlag committed:
Don't persist follow-requests (IOS-192)

zeitschlag committed:
Remove persistence for tags (IOS-192)

Was connected to `let me = authenticationBox.authentication.user(in: managedObjectContext)`

zeitschlag committed:
Use relationships in Report-process (IOS-192)

zeitschlag committed:
Use account for reblogged statuses (stati?) (IOS-192)

zeitschlag committed:
Update header after edit (IOS-192)

zeitschlag committed:
Fix some warnings (IOS-192)

As relationship-information isn't optional anymore, there' no need for ?? everywhere (I just missed some)

zeitschlag committed:
Remove unused code (IOS-192)

zeitschlag committed:
Don't use user but account on Sidebar on iPad (IOS-192)

zeitschlag committed:
Use account to determine status-visibility (IOS-192)

zeitschlag committed:
Replace more users with account (IOS-192)

In Edit and notification

zeitschlag committed:
Remove obsolete wrapper-code

kimar committed:
Hide loading spinner after network requesty

kimar committed:
Remove dead code

zeitschlag committed:
192-profile-about

# Conflicts:
#	Mastodon/Protocol/Provider/DataSourceFacade+Status.swift
#	Mastodon/Scene/Profile/Header/View/ProfileHeaderView+ViewModel.swift
#	Mastodon/Scene/Profile/ProfileViewController.swift
#	MastodonSDK/Sources/MastodonUI/ViewModel/RelationshipViewModel.swift

zeitschlag committed:
Merge branch 'remove_status' into ios-192-profile-about

zeitschlag committed:
Localize button-title for blocked domains (IOS-192)

zeitschlag committed:
Fix warnings and remove unused code (IOS-192)

zeitschlag committed:
Store accounts in container (IOS-192)

In case you see an empty app: Add your account again.

Background: As we need access to the account from the extensions and the extensions can't access the documents-directory but the group-container, well, the accounts will live there from now on.

zeitschlag committed:
Use accounts for widgets (IOS-192)

zeitschlag committed:
Replace another user with an account (IOS-192)

zeitschlag committed:
Merge branch 'develop' into ios-192-profile-about

# Conflicts:
#	Mastodon/Scene/Profile/UserList/FavoritedBy/FavoritedByViewController+DataSourceProvider.swift
#	Mastodon/Scene/Profile/UserList/RebloggedBy/RebloggedByViewController+DataSourceProvider.swift
#	Mastodon/Scene/Profile/UserList/UserListViewModel+Diffable.swift
#	Mastodon/Scene/Profile/UserList/UserListViewModel+State.swift

zeitschlag committed:
Use accounts for shortcut-extension (IOS-192)

This is broken atm, see #1204

zeitschlag committed:
Emojis aren't optional (IOS-192)

meet []

zeitschlag committed:
Use account on account-switcher (IOS-192)

zeitschlag committed:
Fix warnings

zeitschlag committed:
Remove remaining Tag-stuff (IOS-186) (#1205)

zeitschlag committed:
Remove private-note and notification from user (IOS-192)

zeitschlag committed:
Merge branch 'develop' into ios-192-profile-about

# Conflicts:
#	Mastodon.xcodeproj/project.pbxproj
#	Mastodon/Protocol/Provider/DataSourceFacade+Follow.swift
#	Mastodon/Protocol/Provider/DataSourceFacade+Profile.swift
#	Mastodon/Protocol/Provider/DataSourceFacade+Status.swift
#	Mastodon/Protocol/Provider/DataSourceProvider+NotificationTableViewCellDelegate.swift
#	Mastodon/Protocol/Provider/DataSourceProvider+StatusTableViewCellDelegate.swift
#	Mastodon/Protocol/Provider/DataSourceProvider+UITableViewDelegate.swift
#	Mastodon/Protocol/Provider/DataSourceProvider.swift
#	Mastodon/Scene/Discovery/Community/DiscoveryCommunityViewController+DataSourceProvider.swift
#	Mastodon/Scene/Discovery/Community/DiscoveryCommunityViewModel+State.swift
#	Mastodon/Scene/Discovery/Community/DiscoveryCommunityViewModel.swift
#	Mastodon/Scene/Discovery/Posts/DiscoveryPostsViewController+DataSourceProvider.swift
#	Mastodon/Scene/Discovery/Posts/DiscoveryPostsViewModel+State.swift
#	Mastodon/Scene/Discovery/Posts/DiscoveryPostsViewModel.swift
#	Mastodon/Scene/HashtagTimeline/HashtagTimelineViewController+DataSourceProvider.swift
#	Mastodon/Scene/HashtagTimeline/HashtagTimelineViewModel+State.swift
#	Mastodon/Scene/HashtagTimeline/HashtagTimelineViewModel.swift
#	Mastodon/Scene/HomeTimeline/HomeTimelineViewController+DataSourceProvider.swift
#	Mastodon/Scene/HomeTimeline/HomeTimelineViewModel+LoadLatestState.swift
#	Mastodon/Scene/HomeTimeline/HomeTimelineViewModel.swift
#	Mastodon/Scene/Notification/NotificationTimeline/NotificationTimelineViewController+DataSourceProvider.swift
#	Mastodon/Scene/Notification/NotificationTimeline/NotificationTimelineViewController.swift
#	Mastodon/Scene/Notification/NotificationTimeline/NotificationTimelineViewModel.swift
#	Mastodon/Scene/Profile/Bookmark/BookmarkViewController+DataSourceProvider.swift
#	Mastodon/Scene/Profile/Bookmark/BookmarkViewModel+State.swift
#	Mastodon/Scene/Profile/Bookmark/BookmarkViewModel.swift
#	Mastodon/Scene/Profile/Favorite/FavoriteViewController+DataSourceProvider.swift
#	Mastodon/Scene/Profile/Favorite/FavoriteViewModel+State.swift
#	Mastodon/Scene/Profile/Favorite/FavoriteViewModel.swift
#	Mastodon/Scene/Profile/MeProfileViewModel.swift
#	Mastodon/Scene/Profile/ProfileViewController.swift
#	Mastodon/Scene/Profile/ProfileViewModel.swift
#	Mastodon/Scene/Profile/RemoteProfileViewModel.swift
#	Mastodon/Scene/Profile/Timeline/UserTimelineViewController+DataSourceProvider.swift
#	Mastodon/Scene/Profile/Timeline/UserTimelineViewModel+State.swift
#	Mastodon/Scene/Profile/Timeline/UserTimelineViewModel.swift
#	Mastodon/Scene/Report/Report/ReportViewModel.swift
#	Mastodon/Scene/Report/ReportStatus/ReportStatusViewModel+State.swift
#	Mastodon/Scene/Report/ReportStatus/ReportStatusViewModel.swift
#	Mastodon/Scene/Root/MainTab/MainTabBarController.swift
#	Mastodon/Scene/Search/SearchDetail/SearchResult/SearchResultViewController+DataSourceProvider.swift
#	Mastodon/Scene/Search/SearchDetail/SearchResult/SearchResultViewModel+State.swift
#	Mastodon/Scene/Search/SearchDetail/SearchResult/SearchResultViewModel.swift
#	Mastodon/Scene/Share/View/Content/NotificationView+Configuration.swift
#	Mastodon/Scene/Thread/ThreadViewModel.swift
#	MastodonSDK/Sources/MastodonCore/FetchedResultsController/FeedFetchedResultsController.swift
#	MastodonSDK/Sources/MastodonCore/FetchedResultsController/StatusFetchedResultsController.swift
#	MastodonSDK/Sources/MastodonCore/Persistence/FileManager+SearchHistory.swift
#	MastodonSDK/Sources/MastodonCore/Persistence/FileManager+Timeline.swift
#	MastodonSDK/Sources/MastodonCore/Persistence/Persistence.swift
#	MastodonSDK/Sources/MastodonCore/Service/API/APIService+HomeTimeline.swift
#	MastodonSDK/Sources/MastodonCore/Service/API/APIService+Search.swift
#	MastodonSDK/Sources/MastodonSDK/MastodonStatus.swift
#	MastodonSDK/Sources/MastodonUI/Scene/ComposeContent/ComposeContentViewModel.swift
#	MastodonSDK/Sources/MastodonUI/Scene/ComposeContent/Publisher/MastodonStatusPublisher.swift
#	MastodonSDK/Sources/MastodonUI/View/Content/StatusView+Configuration.swift

zeitschlag committed:
Fix warnings

I guess that they made their way back into the app due to the merge

DK1DA commented on Mastodon/Protocol/Provider/DataSourceFacade+Profile.swift:
```diff
@@ -46,54 +46,83 @@ extension DataSourceFacade {
+                                                                                    domain: domain,
+                                                                                    authenticationBox: provider.authContext.mastodonAuthenticationBox) else {
+                    return provider.coordinator.hideLoading()
+                }
```
I know this Is probably a controversial thing, but I really think the following is much more readable (less indentations )

```suggestion
                guard let account = try await provider.context.apiService.fetchUser(
                    username: username,
                    domain: domain,
                    authenticationBox: provider.authContext.mastodonAuthenticationBox
                ) else {
                    return provider.coordinator.hideLoading()
                }
```

DK1DA commented on Mastodon/Protocol/Provider/DataSourceFacade+Profile.swift:
```diff
@@ -46,54 +46,83 @@ extension DataSourceFacade {
+
+                provider.coordinator.hideLoading()
+
+                await coordinateToProfileScene(provider: provider, account: account)
```
Apparently `coordinateToProfileScene(provider:account:)` doesn't need to be awaited for, maybe related to my change on Status?

DK1DA commented on Mastodon/Protocol/Provider/DataSourceFacade+Profile.swift:
```diff
@@ -46,54 +46,83 @@ extension DataSourceFacade {
-                    await coordinateToProfileScene(provider: provider, user: user.asRecord)
-                }
+
+                await coordinateToProfileScene(provider: provider, account: account)
```
Same here regarding `await`, looks like it's not required here

DK1DA commented on Mastodon/Protocol/Provider/DataSourceFacade+Profile.swift:
```diff
@@ -46,54 +46,83 @@ extension DataSourceFacade {
+    static func coordinateToProfileScene(
+        provider: ViewControllerWithDependencies & AuthContextProvider,
+        account: Mastodon.Entity.Account
+    ) {
```
Hoe about making this function `async` ? Then you could get rid of the `Task` below, also the `Task { @MainActor ... ` together with scoping this method `@MainActor` feels a bit weird.

However if you make this method `async` you'll need to keep the `await`s above of course 😄

DK1DA commented on Mastodon/Protocol/Provider/DataSourceFacade+Status.swift:
```diff
@@ -144,18 +144,18 @@ extension DataSourceFacade {
         action: MastodonMenu.Action,
-        menuContext: MenuContext
+        menuContext: MenuContext,
+        completion: ((T) -> Void)? = { (param: Void) in }
```
In other functions, e.g `responseToUserFollowAction` we're simply returning the `Mastodon.Entity.Relationship`, why can't this be done here as well? I feel like another closure makes things more complex than it needs to be.

zeitschlag replied:
Is your question why this method (and the completion) is generic?

Not all menu-actions return a `Mastodon.Entity.Relationship` here. Using the menu you can block a user (returns `Mastodon.Entity.Relationship`), but also a domain returns (`Mastodon.Entity.Empty`) while others like `editStatus` or `translateStatus` do something entirely different.

This was just a first attempt to update a `relationship` IIRC. I consider this WIP and I'm open to other options and ideas!

DK1DA replied:
Ah got you, I thought it might be possible to make `responseToMenuAction` async and to just return `T` instead of having to use a closure. But maybe that's not feasible.

zeitschlag replied:
Let me think about it!

zeitschlag replied:
I thought about it and although this method is already `async`, it can't return anything, as the important stuff happens in `UIAlertAction`-handlers. To my knowledge these handlers don't support return-values. So, if I want something to happen within those handlers, the only way I can think of is a closure.

I also thought about replacing the completion-closure with `.relationshipChanged`-notification. But as the closure already exists and the return-value is used in `DataSourceProvider+NotificationTableViewCellDelegate` to update the menu in `NotificationView`. Sure, we could replace that with a `UIDeferredMenuElement`, but for now I'd prefer to get this PR merged.

DK1DA commented on Mastodon/Protocol/Provider/DataSourceProvider+NotificationTableViewCellDelegate.swift:
```diff
@@ -320,9 +304,11 @@ extension NotificationTableViewCellDelegate where Self: DataSourceProvider & Aut
                 return
             }
 
+            guard let account = notification.status?.entity.account else { return }
```
❓  No directly related to your changes, but I'm wondering if maybe in such cases we should raise an `assertionFailure()` so we're spotting this in Debug builds? (I have also not yet done this, but maybe we should in the future?) WDYT?

DK1DA commented on Mastodon/Protocol/Provider/DataSourceProvider+NotificationTableViewCellDelegate.swift:
```diff
@@ -494,21 +480,18 @@ extension NotificationTableViewCellDelegate where Self: DataSourceProvider & Aut
-                assertionFailure("TODO")
-            default:
-                assertionFailure("TODO")
+                case .status(let status):
```
I'm just wondering about the indentations here because for me, and I think with the Xcode default settings, the above is the correctly indented version.

zeitschlag replied:
Yes, you're right. In some projects, `case` gets indented, in others it's not. I forgot to toggle the switch 🙈

DK1DA commented on Mastodon/Protocol/Provider/DataSourceProvider+StatusTableViewCellDelegate.swift:
```diff
@@ -709,21 +682,21 @@ extension StatusTableViewCellDelegate where Self: DataSourceProvider & AuthConte
-                assertionFailure("TODO")
-            default:
-                assertionFailure("TODO")
+                case .status(let status):
```
Same as my comment regarding indentations above, re-indenting in Xcode it will produce changes here 🤔

DK1DA commented on Mastodon/Scene/Discovery/ForYou/ProfileCardView+Configuration.swift:
```diff
@@ -22,7 +22,7 @@ extension ProfileCardView {
         viewModel.authorAvatarImageURL = account.avatarImageURL()
 
-        let emojis = account.emojis?.asDictionary ?? [:]
+        let emojis = account.emojis.asDictionary
```
😍

DK1DA commented on Mastodon/Scene/Discovery/ForYou/ProfileCardView.swift:
```diff
@@ -316,9 +316,9 @@ extension ProfileCardView {
+            } else if relationship.blocking || relationship.domainBlocking {
                 buttonState = .blocked
-            } else if relationship.requested ?? false {
+            } else if relationship.requested {
```
😍 😍

zeitschlag replied:
I checked with the API and decided to make several properties non-optional

DK1DA commented on Mastodon/Scene/HomeTimeline/HomeTimelineViewModel+LoadLatestState.swift:
```diff
@@ -92,6 +92,7 @@ extension HomeTimelineViewModel.LoadLatestState {
             }
 
             do {
+                await AuthenticationServiceProvider.shared.fetchAccounts(apiService: viewModel.context.apiService)
```
Yeah we have to do something about this. This doesn't feel right. Created [a ticket for this](https://linear.app/mastodon/issue/IOS-216/fix-fetchaccounts-hack).

zeitschlag committed:
Fix indention

Co-authored-by: Marcus Kida <marcus.kida@bearologics.com>

zeitschlag committed:
Fix build (IOS-192)

zeitschlag committed:
Move task and async/await to call-site (IOS-192)

zeitschlag committed:
Fix indention, fix warnings (IOS-192)

DK1DA commented on Mastodon/Protocol/Provider/DataSourceFacade+Profile.swift:
```diff
@@ -41,59 +41,87 @@ extension DataSourceFacade {
-            return
+        provider.coordinator.showLoading()
+
+        Task {
```
This function is already `async` you shouldn't need this additional `Task` here.

DK1DA commented on Mastodon/Protocol/Provider/DataSourceFacade+Profile.swift:
```diff
@@ -41,59 +41,87 @@ extension DataSourceFacade {
-        guard let domain = account.domain else { return provider.coordinator.hideLoading() }
-        
+
         Task {
```
I think this `Task` can also be removed.

DK1DA commented on Mastodon/Scene/Onboarding/Login/MastodonLoginViewController.swift:
```diff
@@ -139,30 +139,23 @@ class MastodonLoginViewController: UIViewController, NeedsDependency {
+                        _ = try await self.context.authenticationService.activeMastodonUser(domain: domain, userID: account.id)
+                        FileManager.default.store(account: account, forUserID: MastodonUserIdentifier(domain: domain, userID: account.id))
+                        Task { @MainActor in
+                            self.coordinator.setup()
```
This feels a bit weird, nesting a `Task` inside a Combine sink, then spawning yet another `Task`.
I actually really like the `asyncMap` approach, and maybe we could annotate `coordinator.setup()` with `@MainActor` so we can get rid of the second `Task` her?

DK1DA commented on Mastodon/Scene/Profile/Header/View/ProfileHeaderView.swift:
```diff
@@ -238,20 +235,14 @@ final class ProfileHeaderView: UIView {
-}
+    init(account: Mastodon.Entity.Account, me: Mastodon.Entity.Account, relationship: Mastodon.Entity.Relationship?) {
+
+        viewModel = ViewModel(account: account, me: me, relationship: relationship)
```
I would probably construct the `ProfileHeaderView` with the ViewModel here to only have the three params in the ViewModel's constructor.

zeitschlag replied:
I keep thinking: Why not just using the `ProfileViewModel` or use something similar to your `MastodonStatus`?

DK1DA commented on Mastodon/Scene/Profile/ProfileViewController.swift:
```diff
@@ -202,33 +199,38 @@ extension ProfileViewController {
 
             var items: [UIBarButtonItem] = []
             defer {
-                self.navigationItem.rightBarButtonItems = !items.isEmpty ? items : nil
```
🥲 why not just

```
self.navigationItem.rightBarButtonItems = items.isEmpty ? nil : items
```

Isn't this short and concise enough, and maybe even get rid of the `items.isNotEmpty` extension.

DK1DA commented on Mastodon/Scene/Profile/ProfileViewController.swift:
```diff
@@ -202,33 +199,38 @@ extension ProfileViewController {
             }
 
-            guard !isSuspended else {
+            if let suspended = self.viewModel.account.suspended, suspended == true {
```
Maybe

```suggestion
            guard self.viewModel.account.suspended == true else {
```

to have it uniform with the other `guard`?

DK1DA commented on Mastodon/Scene/Profile/ProfileViewController.swift:
```diff
@@ -787,101 +754,128 @@ extension ProfileViewController: ProfileHeaderViewControllerDelegate {
         }
-        
+
+        //            switch relationshipAction {
```
Just a reminder to remove this code.

DK1DA commented on MastodonSDK/Sources/MastodonCore/Persistence/FileManager+Shared.swift:
```diff
@@ -0,0 +1,18 @@
+    }
+
+    var sharedDirectory: URL? {
+        containerURL(forSecurityApplicationGroupIdentifier: AppName.groupID)
```
👍

DK1DA commented on MastodonSDK/Sources/MastodonCore/Persistence/FileManager+Timeline.swift:
```diff
@@ -1,22 +1,21 @@
-import MastodonCore
 import MastodonSDK
 
 extension FileManager {
```
You could also just make the extension public, then all function will implicitly inherit the access modifier.

DK1DA commented on MastodonSDK/Sources/MastodonSDK/Entity/Mastodon+Entity+Account.swift:
```diff
@@ -36,55 +35,85 @@ extension Mastodon.Entity {
-            case suspended
-            case muteExpiresAt = "mute_expires_at"
+
+        internal init(id: Mastodon.Entity.Account.ID, username: String, acct: String, url: String, displayName: String, note: String, avatar: String, avatarStatic: String? = nil, header: String, headerStatic: String? = nil, locked: Bool, emojis: [Mastodon.Entity.Emoji] = [], discoverable: Bool? = nil, createdAt: Date, lastStatusAt: Date? = nil, statusesCount: Int, followersCount: Int, followingCo ...
```
😱

zeitschlag replied:
Ah! I needed that in the beginning for a `placeholder`, but forgot to remove it afterwards. Thank you!

zeitschlag committed:
Remove unused code

zeitschlag committed:
Split private/public into their own extension

DK1DA commented on Mastodon/Scene/Account/AccountListViewModel.swift:
```diff
@@ -103,25 +103,25 @@ extension AccountListViewModel {
     ) {
-        guard let user = authentication.user(in: context) else { return }
-        
+        guard let account = authentication.account() else { return }
```
For me this method returns `nil` thus bailing the user out here and leaving the app in a dangling (everything is blank) state.

Steps to reproduce (at least for me):
* Install `develop` branch version
* Log in (with multiple accounts in my case)
* Install this PR branch

zeitschlag replied:
I will look into that, thank you!

Great catch, I can reproduce it, too and I think, I found the culprit: The renaming of `FeedFetchedResultsController` causes this as there was a [`fetchAccounts`](https://github.com/mastodon/mastodon-ios/pull/1192/commits/febbc6f22abc7e62545d16a3e5a29ec463920398#diff-ae17b729d2c09d8dfaffa6380bfabaa33e364f9eb5c0dd48f0a9a655bffdfc47L80-L95) when loading the home-timeline.

I'll look into it how to fix that.

Another way to reproduce this is to remove the account-JSON-files in the shared directory.

zeitschlag replied:
Should work again :)

DK1DA replied:
Yes works now, thank you 👍

zeitschlag committed:
Reduce nested-ness of tasks (IOS-192)

Thank you @kimar

zeitschlag committed:
Use `guard` (IOS-192)

zeitschlag committed:
Download accounts at start and don't fail if there's no me (IOS-192)

Known issue for know is that the profile-tab shows a blank screen. Still gotta fix that.

zeitschlag committed:
Refactor tab/MainTabBarController to add viewcontrollers as properties

This is a WIP-step for account-stuff. Also: iPhone only, iPad should come next

zeitschlag committed:
Fix profile-button (IOS-192)

zeitschlag committed:
[WIP] Add relationships/user to notifications (IOS-192)

zeitschlag committed:
[WIP] Preparation to get relationship into menu on notification-screen (IOS-192)

zeitschlag committed:
Remove unused property (IOS-192)

zeitschlag committed:
Remove code (IOS-192)

zeitschlag committed:
Group files and use simpler context (IOS-192)

zeitschlag committed:
Don't use publisher for most properties in NotificationView (IOS-192)

Timestamp and A11y are still missing (and A11y is broken atm)

zeitschlag committed:
Remove unused code (IOS-192)

zeitschlag committed:
Move notification-timestamp away from viewmodel (IOS-192)

zeitschlag committed:
Set a11y-label for notification-screen (IOS-192)

zeitschlag committed:
Move over a11y-actions for notifications (IOS-192)

zeitschlag committed:
Make follow-requests work again (IOS-192)

Aaaaand remove obsolete viewmodel and put relationship to make menu work again (too)

zeitschlag committed:
Merge branch 'develop' into ios-192-profile-about

# Conflicts:
#	Mastodon/Protocol/Provider/DataSourceProvider+NotificationTableViewCellDelegate.swift
#	Mastodon/Scene/Notification/NotificationTimeline/NotificationTimelineViewController+DataSourceProvider.swift
#	MastodonSDK/Sources/MastodonCore/DataController/FeedDataController.swift
#	MastodonSDK/Sources/MastodonUI/View/Content/StatusView+Configuration.swift

zeitschlag committed:
Revert "Use `guard` (IOS-192)"

This reverts commit 64dc97ab9bb50041146777248885a67dc1fcf73b.

zeitschlag committed:
Enable domain-block for User-refactoring (IOS-192)

zeitschlag committed:
Merge branch 'develop' into ios-192-profile-about

zeitschlag committed:
Merge branch 'develop' into ios-192-profile-about

zeitschlag committed:
Merge branch 'develop' into ios-192-profile-about

# Conflicts:
#	MastodonSDK/Sources/MastodonUI/View/Control/ProfileRelationshipActionButton.swift

zeitschlag committed:
Remove obsolete code (IOS-192)

We didn't use that as someone rebuilt this screen in SwiftUI.

zeitschlag committed:
Remove more users (IOS-192)

zeitschlag committed:
Fix another user-warning (IOS-192)

zeitschlag committed:
Update account-screen when following/unfollowing/... (IOS-192)

zeitschlag committed:
Remove persistence for user (IOS-192)

zeitschlag committed:
Remove more users (IOS-192)

zeitschlag committed:
Remove shadow

zeitschlag committed:
Remove obsolete code (IOS-192)

zeitschlag committed:
Make the app work on iPad again (IOS-192)

Friends don't let friends use forced unwrapping.

DK1DA commented on Mastodon/Coordinator/SceneCoordinator.swift:
```diff
@@ -97,24 +98,53 @@ final public class SceneCoordinator {
                                     case .follow:
-                                        let profileViewModel = RemoteProfileViewModel(context: appContext, authContext: authContext, notificationID: notificationID)
-                                        _ = self.present(scene: .profile(viewModel: profileViewModel), from: from, transition: .show)
+                                        Task {
```
This feels a bit weird, as we've got a 

```swift
Task {
  DispatchQueue.main.async {
    Task {
      ...
    }
  }
}
```

here, could we maybe eliminate the GCD dispatch and leave it with one task? Maybe some `@MainActor`scoped function used inside etc?

Ref: https://github.com/mastodon/mastodon-ios/pull/1192/files#diff-a97a67b1084ce12242b4ac094ec7976f865dac1a8a18c99323b5fa99d6834981L54

and: https://github.com/mastodon/mastodon-ios/pull/1192/files#diff-a97a67b1084ce12242b4ac094ec7976f865dac1a8a18c99323b5fa99d6834981L78

zeitschlag replied:
Good catch!

zeitschlag committed:
Update accounts on profile-screen on pull to refresh (IOS-192)

zeitschlag committed:
Remove unused code (IOS-192)

zeitschlag committed:
Update screen when relationship changes on another screen (IOS-192)

zeitschlag committed:
FIx warning

zeitschlag committed:
Update profile-screen when relationship changes on user list (IOS-192)

zeitschlag committed:
Slighty refactor relationship-update-notification-handling (IOS-192)

Also: consider menu-changes from profile-screen

zeitschlag committed:
Rework relationship-notification (IOS-192)

They get sent from the DataSourceFacade now

DK1DA commented on Mastodon/Protocol/Provider/DataSourceFacade+Block.swift:
```diff
@@ -13,56 +13,40 @@ import MastodonSDK
-        let apiService = dependency.context.apiService
-        let authBox = dependency.authContext.mastodonAuthenticationBox
+        let userInfo = [
+            "relationship": response.value,
```
Should we use a string enum here to prevent issues by typos?

DK1DA commented on Mastodon/Protocol/Provider/DataSourceFacade+Follow.swift:
```diff
@@ -7,40 +7,29 @@
         dependency.context.authenticationService.fetchFollowingAndBlockedAsync()
 
+        NotificationCenter.default.post(name: .relationshipChanged, object: nil, userInfo: [
+            "relationship": response
```
Same question about using a string enum here

zeitschlag committed:
Merge branch 'develop' into ios-192-profile-about

DK1DA commented on Mastodon/Scene/MediaPreview/MediaPreviewViewModel.swift:
```diff
@@ -151,8 +151,8 @@ extension MediaPreviewViewModel {
-                guard !assertURL.hasSuffix("missing.png") else { return false }
-                return true
+
+                return assertURL.hasSuffix("missing.png") == false
```
Maybe move this to a string constant somewhere so we can adjust in case the name of this file changes.

DK1DA commented on Mastodon/Scene/Notification/NotificationView/NotificationView+Configuration.swift:
```diff
@@ -0,0 +1,281 @@
+            self.rejectFollowRequestButton.tintColor = .clear
+            self.rejectFollowRequestButton.setTitleColor(.clear, for: .normal)
+        } else {
+            self.rejectFollowRequestActivityIndicatorView.stopAnimating()
```
If `state == .isAccepting` then this code will be executed.
I would use a `switch` (or if/else) here otherwise this code is hard to interpret.

zeitschlag replied:
From what I understand, as I just [copied the code](https://github.com/mastodon/mastodon-ios/commit/7dfa56507e5e2accd7ce8714e748bd159732e7e9#diff-1c984bd601e5b7ddced8d31f5b1b4a98a9e0557b91ec0ee1257b8b78d8a4f745) over:

If `state == .isAccepting`, in fact, if `state` is something else than `isRejecting`, we want the `rejectFollowRequestActivityIndicatorView` to stop spinning (vice versa for the `acceptFollowRequestButton`). Using a `switch` for that would result in lots of code duplication, but maybe I'm wrong?

So yes, there might be more elegant ways, I just didn't want to break anything here :)

DK1DA replied:
Ah, fair enough! Thanks for the clarification, I just re-read the code and it makes sense to me now 👍

DK1DA commented on Mastodon/Scene/Profile/Header/ProfileHeaderViewController.swift:
```diff
@@ -81,14 +82,30 @@ final class ProfileHeaderViewController: UIViewController, NeedsDependency, Medi
         super.viewDidLoad()
-        
+
+
```
Excessive blank line here

zeitschlag committed:
Use constant foe UserInfo-key instead of string (IOS-192)

DK1DA commented on Mastodon/Scene/Root/Tab.swift:
```diff
@@ -0,0 +1,111 @@
+        return image.withRenderingMode(.alwaysTemplate).resized(size: CGSize(width: 80, height: 80))
+    }
+
+//    @MainActor
```
Reminder: Remove this code.

DK1DA commented on Mastodon/Scene/Notification/NotificationViewController.swift:
```diff
@@ -21,7 +21,7 @@ final class NotificationViewController: TabmanViewController, NeedsDependency {
     var observations = Set<NSKeyValueObservation>()
 
-    var viewModel: NotificationViewModel!
+    var viewModel: NotificationViewModel?
```
I'm not a huge fan of having those optional view models, because the view itself isn't functional without the VMs. But also there is probably no quick way of changing this, from an architectural perspective.

zeitschlag replied:
I'm not a huge fan of forced unwrapping, it feels like a very dirty hack throughout the entire app just to be able to use ViewModels literally everywhere.

But: `?` has a huge advantage over `!`: It doesn't crash, so I consider this a quick workaround.

DK1DA commented on Mastodon/Scene/Share/View/Content/UserView+Configuration.swift:
```diff
@@ -17,68 +17,16 @@ import MastodonSDK
 
         do {
-            let emojis = account.emojis?.asDictionary ?? [:]
+            let emojis = account.emojis.asDictionary
```
❤️

DK1DA commented on MastodonSDK/Sources/MastodonCore/AuthenticationServiceProvider.swift:
```diff
@@ -101,10 +101,29 @@ public extension AuthenticationServiceProvider {
+
+        NotificationCenter.default.post(name: .userFetched, object: nil)
+
+        completion?()
```
Maybe we can get rid of this closure and use the `.userFetched`notification instead?

zeitschlag replied:
We can get rid of `completion` as we don't use that anymore. Good catch 👍

kimar reviewed (approved):
LGTM Code Wise, nothing major, with the exception of [this comment](https://github.com/mastodon/mastodon-ios/pull/1192/files#r1494206691) that might actually lead to issues. Apart from that, good work, going to give it a thorough test now.

zeitschlag committed:
Remove unused code (IOS-192)

zeitschlag committed:
Fix build

zeitschlag committed:
Use constant (IOS-192)

zeitschlag committed:
Remove unused completion (IOS-192)

zeitschlag committed:
Recude task-nestedness (IOS-192)

zeitschlag committed:
Don't reload content all the time (IOS-192)

zeitschlag committed:
Fix warning

zeitschlag committed:
Merge branch 'develop' into ios-192-profile-about

zeitschlag committed:
Show loading-indicator (IOS-192)

zeitschlag committed:
Remove nested tasks (IOS-192)

zeitschlag committed:
Cleanup (IOS-192)

zeitschlag committed:
Merge branch 'develop' into ios-192-profile-about

kimar reviewed (approved):
LGTM 🚀  Great work!
